File: cpp/BOJ/BOJ1759.cc
#include <iostream>
#include <map>
#include <vector>
#include <string>
#include <algorithm>
#include <set>
using namespace std;

int L,C;
map<char,int> isVowels;
vector<char> alphabets;
vector<char> candidates;
vector<bool> isUsed;
set<string> answer;

void combination(int idx, int pick)
{
    if (pick == L)
    {
        //isAnswer
        int vowelCnt = 0;
        int consonantCnt = 0;
        
        for (char c : candidates)
        {
            if (isVowels.find(c) == isVowels.end())
            {
                //자음
                consonantCnt += 1;
            }
            else
            {
                //모음
                vowelCnt += 1;
            }
        }

        if (vowelCnt >= 1 && consonantCnt >= 2)
        {
            string str = "";
            for (char c : candidates)
            {
                str += c;
            }

            sort(str.begin(), str.end());
            answer.insert(str);
        }

        return;
    }

    if (idx == C)
    {
        return;
    }

    if (!isUsed[idx])
    {
        isUsed[idx] = true;
        candidates.push_back(alphabets[idx]);
        combination(idx+1,pick+1);
        candidates.pop_back();
        isUsed[idx] = false;
    }
    
    combination(idx+1,pick);
    return;
}

int main(void) {

    isVowels.insert({'a',1});
    isVowels.insert({'e',1});
    isVowels.insert({'i',1});
    isVowels.insert({'o',1});
    isVowels.insert({'u',1});

    cin >> L >> C;

    isUsed.resize(C,false);

    for (int loop = 0; loop < C; ++loop)
    {
        char c;
        cin >> c;
        alphabets.push_back(c);
    }    

    combination(0,0);
    
    for (string ans : answer)
    {
        cout << ans << "\n";
    }

    return 0;
}

File: java/BOJ/BOJ1759.java
package BOJ;

import java.io.*;
import java.util.*;

public class BOJ1759 {
	
	static int L,C;
	static Set<Character> vowels = new LinkedHashSet<>();
	static String origin;
	static char[] candidate;
	static boolean[] isSelected;
	static List<String> answers = new LinkedList<>();
	
	public static boolean isAnswer() {
		int consonantCnt, vowCnt = 0;
		
		for (char alphabet : candidate)
		{
			if (vowels.contains(alphabet))
				vowCnt += 1;
		}

		consonantCnt = candidate.length - vowCnt;
		
		return (consonantCnt >= 2) && (vowCnt >= 1);
	}
	
	public static void combination(int idx, int pickCnt) {
	
		if (pickCnt == L)
		{
			if (isAnswer()) {
				String answer = String.valueOf(candidate);
				answers.add(answer);
			}
			
			return;
		}
		else if (idx == C)
		{
			return;
		}
		
		int alphabetIdx = (int)origin.charAt(idx) - (int)'a';	
		
		if (!isSelected[alphabetIdx])
		{
			isSelected[alphabetIdx] = true;
			candidate[pickCnt] = origin.charAt(idx);
			combination(idx+1,pickCnt+1);
			isSelected[alphabetIdx] = false;
		}
		
		combination(idx + 1, pickCnt);
	}
	
	public static void main(String[] args) throws IOException {
		
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		
		String[] alphabetNumbers = br.readLine().split(" ");
		String[] alphabets = br.readLine().split(" ");
		Arrays.sort(alphabets);
		
		L = Integer.parseInt(alphabetNumbers[0]);
		C = Integer.parseInt(alphabetNumbers[1]);
	
		StringBuilder sb = new StringBuilder();

		for (String alphabet : alphabets)
		{
			sb.append(alphabet);
		}
		
		origin = sb.toString();
		
		isSelected = new boolean[26];
		candidate = new char[L];
		
		vowels.add('a');
		vowels.add('e');
		vowels.add('i');
		vowels.add('o');
		vowels.add('u');
		
		combination(0,0);
		
		for (String answer : answers)
		{
			bw.append(answer);
			bw.newLine();
		}
		
		bw.flush();
	}
}
